File: symmetric-encryption/js/encrypt.js
#!/usr/bin/env node

const fs = require('fs');
const crypto = require('crypto');

// Read the hex-encoded secret key from the first command-line argument.
const secret = Buffer.from(process.argv[2], 'hex');

// Generate a random IV.
const iv = crypto.randomBytes(16);

// Read the data to be encrypted from stdin.
const info = fs.readFileSync(0, 'utf-8');

// Encrypt the data with AES.
const cipher = crypto.createCipheriv('aes256', secret, iv);
const encrypted = cipher.update(info, 'binary', 'base64') + cipher.final('base64');
const encodedIV = Buffer.from(iv).toString('base64');

// Concatenate the encoded ciphertext, '--', and the IV, then encode as Base64.
data = Buffer.from(`${ encrypted }--${ encodedIV }`).toString('base64');

// Calculate the digest.
const digest = crypto.createHmac('sha1', secret).update(data).digest('hex');

// Concatenate the data and the digest then output it.
console.log(`${data}--${digest}`);

File: symmetric-encryption/python/encrypt.py
#!/usr/bin/env python3

import sys
import hmac
import base64
import binascii

from Cryptodome.Cipher import AES
from Cryptodome.Util.Padding import pad

if __name__ == '__main__':
	# Read the hex-encoded secret key from the first command-line argument.
	key = binascii.unhexlify(sys.argv[1])

	# Read the data to be encrypted from stdin.
	plaintext = sys.stdin.buffer.read()

	# Encrypt the data with AES.
	cipher = AES.new(key, AES.MODE_CBC)
	ct_bytes = cipher.encrypt(pad(plaintext, AES.block_size))
	ciphertext = base64.b64encode(ct_bytes)
	iv = base64.b64encode(cipher.iv)

	# Concatenate the encoded ciphertext, '--', and the IV, then encode as Base64.
	data = base64.b64encode(ciphertext + b'--' + iv)

	# Calculate the digest.
	digest = binascii.hexlify(hmac.digest(key, data, 'sha1'))

	# Concatenate the data and the digest then output it.
	print((data + b'--' + digest).decode('utf-8'))
